temporian/implementation/numpy/operators/set_index.py

```diff
@@ -0,0 +1,186 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import numpy as np
+
+from temporian.core.operators.set_index import SetIndexOperator
+from temporian.implementation.numpy.data.event import NumpyEvent
+from temporian.implementation.numpy.data.event import NumpyFeature
+from temporian.implementation.numpy.data.sampling import NumpySampling
+from temporian.implementation.numpy.operators.utils import _sort_by_timestamp
+
+
+def _append_impl(event: NumpyEvent, append_feat_names: List[str]) -> NumpyEvent:
+    """
+    Append the specified feature names to the index of the given NumpyEvent.
+
+    Args:
+        event:
+            The input NumpyEvent object.
+        append_feat_names:
+            A list of feature names to append to the index.
+
+    Returns:
+        The resulting NumpyEvent object with the updated index.
+    """
+    # destination index names
+    dst_idx_names = event.sampling.index + append_feat_names
+
+    # positions of features that are going to be part of the destination index
+    dst_idx_pos = [
+        event.feature_names.index(append_feat_name)
+        for append_feat_name in append_feat_names
+    ]
+    # positions of features that are being kept
+    dst_feat_pos = {
+        feat_name: pos
+        for pos, feat_name in enumerate(event.feature_names)
+        if feat_name not in append_feat_names
+    }
+    # initialize destination event & sampling data
+    dst_event_data = {}
+    dst_samp_data = {}
+    for src_idx_lvl, src_idx_lvl_feats in event.data.items():
+        dst_idx_suffs = [
+            tuple(x)
+            for x in zip(*[src_idx_lvl_feats[i].data for i in dst_idx_pos])
+        ]
+        for target_dst_idx_suff in set(dst_idx_suffs):
+            # find all occurrences of destination suff in source event
+            dst_idx_suff_pos = [
+                idx
+                for idx, dst_idx_suff in enumerate(dst_idx_suffs)
+                if dst_idx_suff == target_dst_idx_suff
+            ]
+            # create destination index
+            dst_idx_lvl = src_idx_lvl + tuple(target_dst_idx_suff)
+
+            dst_event_data[dst_idx_lvl] = [
+                NumpyFeature(
+                    feat_name,
+                    event.data[src_idx_lvl][feat_pos].data[dst_idx_suff_pos],
+                )
+                for feat_name, feat_pos in dst_feat_pos.items()
+            ]
+            # fill sampling data
+            dst_samp_data[dst_idx_lvl] = event.sampling.data[src_idx_lvl][
+                dst_idx_suff_pos
+            ]
+
+    # finally, sort according to timestamps. TODO: this is merging sorted
+    # arrays, we should later improve this code by avoiding the full sort
+    _sort_by_timestamp(dst_event_data, dst_samp_data)
```

Comment:
Might be wrong, but won't this be sorted already? Since we're taking subarrays of already sorted arrays

Reply:
not necessarily - they would also need to be sorted suffix-level-wise for that to happen